// typia/lib/transform.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.transform = void 0;
var typescript_1 = __importDefault(require("typescript"));
var FileTransformer_1 = require("./transformers/FileTransformer");
var transform = function (program, options) {
    var compilerOptions = program.getCompilerOptions();
    var strict = compilerOptions.strictNullChecks !== undefined
        ? !!compilerOptions.strictNullChecks
        : !!compilerOptions.strict;
    if (strict === false)
        throw new Error("Error on \"tsconfig.json\": typia requires `compilerOptions.strictNullChecks` to be true.");
    return FileTransformer_1.FileTransformer.transform({
        program: program,
        compilerOptions: compilerOptions,
        checker: program.getTypeChecker(),
        printer: typescript_1.default.createPrinter(),
        options: options || {},
    });
};
exports.transform = transform;
exports.default = exports.transform;
