// typia/lib/metadata/MetadataArray.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetadataArray = void 0;
var MetadataArray = (function () {
    function MetadataArray(props) {
        this.name = props.name;
        this.value = props.value;
        this.index = props.index;
        this.recursive = props.recursive;
        this.nullables = props.nullables;
    }
    MetadataArray._From_without_value = function (props) {
        return this.create({
            name: props.name,
            value: null,
            index: props.index,
            recursive: props.recursive,
            nullables: props.nullables,
        });
    };
    MetadataArray.create = function (props) {
        return new MetadataArray(props);
    };
    MetadataArray.prototype.toJSON = function () {
        return {
            name: this.name,
            value: this.value.toJSON(),
            nullables: this.nullables,
            recursive: this.recursive,
            index: this.index,
        };
    };
    return MetadataArray;
}());
exports.MetadataArray = MetadataArray;
