// typia/lib/metadata/MetadataTuple.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetadataTuple = void 0;
var MetadataTuple = (function () {
    function MetadataTuple(props) {
        this.name = props.name;
        this.elements = props.elements;
        this.index = props.index;
        this.recursive = props.recursive;
        this.nullables = props.nullables;
    }
    MetadataTuple._From_without_elements = function (props) {
        return this.create({
            name: props.name,
            index: props.index,
            elements: null,
            recursive: props.recursive,
            nullables: props.nullables.slice(),
        });
    };
    MetadataTuple.create = function (props) {
        return new MetadataTuple(props);
    };
    MetadataTuple.prototype.toJSON = function () {
        return {
            name: this.name,
            index: this.index,
            elements: this.elements.map(function (elem) { return elem.toJSON(); }),
            recursive: this.recursive,
            nullables: this.nullables.slice(),
        };
    };
    return MetadataTuple;
}());
exports.MetadataTuple = MetadataTuple;
