// @blockchain-lab-um/did-provider-key/dist/keyDidResolver.js
import { decodePublicKey } from '@blockchain-lab-um/utils';
import { getResolver } from '@cef-ebsi/key-did-resolver';
import { Resolver, } from 'did-resolver';
import { curveResolverMap } from './curves.js';
import { checkDidComponents, getKeyType } from './keyDidUtils.js';
export const resolveDid = (did) => {
    const components = checkDidComponents(did);
    const publicKey = decodePublicKey(components.multibaseValue);
    const keyType = getKeyType(publicKey.code);
    if (!curveResolverMap[keyType]) {
        throw new Error('invalidDid: invalid key type');
    }
    return curveResolverMap[keyType]({
        didIdentifier: components.multibaseValue,
        publicKey,
        keyType,
    });
};
export const resolveSecp256k1Ebsi = async (did) => {
    const keyResolver = getResolver();
    const didResolver = new Resolver(keyResolver);
    const resolution = await didResolver.resolve(did);
    return resolution.didDocument;
};
const startsWithMap = {
    'did:key:zQ3s': resolveDid,
    'did:key:zDn': resolveDid,
    'did:key:z6Mk': resolveDid,
    'did:key:z2dm': resolveSecp256k1Ebsi,
    'did:key:zBhB': resolveSecp256k1Ebsi,
};
export const resolveDidKey = async (didUrl, parsed, _resolver, _options) => {
    try {
        const startsWith = parsed.did.substring(0, 12);
        if (startsWithMap[startsWith] !== undefined) {
            const didDocument = await startsWithMap[startsWith](didUrl);
            return {
                didDocumentMetadata: {},
                didResolutionMetadata: {},
                didDocument,
            };
        }
        return {
            didDocumentMetadata: {},
            didResolutionMetadata: {
                error: 'invalidDid',
                message: 'unsupported key type for did:key',
            },
            didDocument: null,
        };
    }
    catch (err) {
        return {
            didDocumentMetadata: {},
            didResolutionMetadata: {
                error: 'invalidDid',
                message: err.toString(),
            },
            didDocument: null,
        };
    }
};
export function getDidKeyResolver() {
    return { key: resolveDidKey };
}
