// @blockchain-lab-um/did-provider-key/dist/keyDidProvider.js
import { createJWK, encodePublicKey, getCompressedPublicKey, } from '@blockchain-lab-um/utils';
import { util } from '@cef-ebsi/key-did-resolver';
import { AbstractIdentifierProvider } from '@veramo/did-manager';
import { isSupportedKeyType, KEY_TYPE_TO_MULTICODEC_NAME, } from './types/keyDidTypes.js';
export class KeyDIDProvider extends AbstractIdentifierProvider {
    constructor(options) {
        super();
        this.defaultKms = options.defaultKms;
    }
    async createIdentifier({ kms, options }, context) {
        const keyType = options?.keyType && isSupportedKeyType(options.keyType)
            ? options.keyType
            : 'Ed25519';
        const key = await this.importOrGenerateKey({
            kms: kms ?? this.defaultKms,
            options: {
                keyType,
                ...(options?.privateKeyHex && {
                    privateKeyHex: options.privateKeyHex,
                }),
            },
        }, context);
        if (options?.type === 'ebsi') {
            const compressedKey = keyType === 'Secp256k1'
                ? getCompressedPublicKey(`0x${key.publicKeyHex}`)
                : key.publicKeyHex;
            const jwk = createJWK(keyType, compressedKey);
            const did = util.createDid(jwk);
            const identifier = {
                did,
                controllerKeyId: key.kid,
                keys: [key],
                services: [],
            };
            return identifier;
        }
        const compressedKey = keyType === 'Secp256k1'
            ? getCompressedPublicKey(`0x${key.publicKeyHex}`)
            : key.publicKeyHex;
        const methodSpecificId = encodePublicKey(Buffer.from(compressedKey, 'hex'), KEY_TYPE_TO_MULTICODEC_NAME[keyType]);
        const identifier = {
            did: `did:key:${methodSpecificId}`,
            controllerKeyId: key.kid,
            keys: [key],
            services: [],
        };
        return identifier;
    }
    async updateIdentifier(_args, _context) {
        throw new Error('KeyDIDProvider updateIdentifier not supported yet.');
    }
    async deleteIdentifier(identifier, context) {
        for (const { kid } of identifier.keys) {
            await context.agent.keyManagerDelete({ kid });
        }
        return true;
    }
    async addKey(_args, _context) {
        throw Error('KeyDIDProvider addKey not supported');
    }
    async addService(_args, _context) {
        throw Error('KeyDIDProvider addService not supported');
    }
    async removeKey(_args, _context) {
        throw Error('KeyDIDProvider removeKey not supported');
    }
    async removeService(_args, _context) {
        throw Error('KeyDIDProvider removeService not supported');
    }
    async importOrGenerateKey(args, context) {
        if (args.options.privateKeyHex) {
            return context.agent.keyManagerImport({
                kms: args.kms || this.defaultKms,
                type: args.options.keyType,
                privateKeyHex: args.options.privateKeyHex,
            });
        }
        return context.agent.keyManagerCreate({
            kms: args.kms || this.defaultKms,
            type: args.options.keyType,
        });
    }
}
