// @blockchain-lab-um/veramo-datamanager/dist/data-store/memoryDataStore.js
import jsonpath from 'jsonpath';
import { v4 } from 'uuid';
import { AbstractDataStore, } from './abstractDataStore.js';
export class MemoryDataStore extends AbstractDataStore {
    constructor() {
        super(...arguments);
        this.data = {};
    }
    async save(args) {
        const id = v4();
        this.data[id] = args.data;
        return id;
    }
    async delete(args) {
        const { id } = args;
        if (id in this.data) {
            delete this.data[id];
            return true;
        }
        return false;
    }
    async query(args) {
        const { filter } = args;
        if (filter && filter.type === 'id') {
            try {
                if (this.data[filter.filter]) {
                    const obj = [
                        {
                            metadata: { id: filter.filter },
                            data: this.data[filter.filter],
                        },
                    ];
                    return obj;
                }
                return [];
            }
            catch (e) {
                throw new Error('Invalid id');
            }
        }
        if (filter === undefined || (filter && filter.type === 'none')) {
            return Object.keys(this.data).map((k) => ({
                metadata: { id: k },
                data: this.data[k],
            }));
        }
        if (filter && filter.type === 'JSONPath') {
            const objects = Object.keys(this.data).map((k) => ({
                metadata: { id: k },
                data: this.data[k],
            }));
            const filteredObjects = jsonpath.query(objects, filter.filter);
            return filteredObjects;
        }
        return [];
    }
    async clear(args) {
        this.data = {};
        return true;
    }
}
